### DISCUSSION TITLE
How do you start the prod server using Docker and SQLite?

### DISCUSSION THREAD

**Discussion by daniel-walz** [adonisjs/core#2171] [Help] [answered]
I have a question on how people setup their AdonisJS apps with Docker in production. When running locally I just set volumes that point to my local folders so that I can develop without rebuilding and restarting the image and container. And copy over my .env file to the container.

In production I'm a little confused on how to handle the .env file and then the DB migration step. I assume that I should setup a secret value in my workflow for the app key and send that to Docker using -e APP_KEY={{my_secret_ap_key}}.

`docker run -p 3333:3333 -e PORT=3333 -e HOST=0.0.0.0 -e APP_KEY={{secret_app_key}} -e NODE_ENV=production -e DB_CONNECTION=sqlite -t settings-service:latest
`

When I run the container, it seems like it starts and migration runs successfully, but the application spits out an error when I try to access the DB:

```
❯ migrated database/migrations/1611240510677_settings

Migrated in 73 ms

> settings-service@1.0.0 start /app
> node .

{"level":30,"time":1611568765505,"pid":31,"hostname":"7c724bf9d0da","name":"settings-service","msg":"started server on 0.0.0.0:3333"}
[Error: SQLITE_CANTOPEN: unable to open database file] {
  errno: 14,
  code: 'SQLITE_CANTOPEN'
}
```

Dockerfile
```
# base image for local development
FROM mhart/alpine-node:14

# SQLite3
RUN apk update
RUN apk add git sqlite python make g++ > /dev/null

# set working directory
WORKDIR /app

# add environment variables
ENV PATH /app/node_modules/.bin:$PATH
ENV SKIP_PREFLIGHT_CHECK=true

# copy application files
COPY . .

# create SQLite DB directory
RUN mkdir -p tmp

# install and cache app dependencies
RUN npm config set registry http://registry.npmjs.org/
RUN npm install --silent
RUN npm run build

RUN chmod +x start.sh

CMD ["/app/start.sh"]
```

start.sh
```
#!/bin/sh

node ace migration:run --force
npm run start
```

Note, that in this case I have already run the invoke command for Lucid and checked in everything it generates and sets up.
`node ace invoke @adonisjs/lucid`

This was the only way I got the migration to run in the container.
I have also verified in that the DB table is created when I exec into the container.

```
/app/tmp # sqlite3
SQLite version 3.30.1 2019-10-10 20:19:45
Enter ".help" for usage hints.
Connected to a transient in-memory database.
Use ".open FILENAME" to reopen on a persistent database.
sqlite> .open db.sqlite3
sqlite> .tables
adonis_schema  settings
sqlite> SELECT * FROM adonis_schema
   ...> ;
1|database/migrations/1611240510677_settings|1|2021-01-25 09:59:24
```

What am I missing?

**Comment by daniel-walz** [comment, marked as answer]
I think I found the answer, I was setting up the tmp folder for the SQLite DB in the working dir (/app) in the container, not in the build folder from where I start the application after I have done a production build.

The fix was to move the creation of the tmp folder, remove the line inte the Dockerfile and add it in the start script.

start.sh
```
#!/bin/sh

cd build
mkdir -p tmp
node ace migration:run --force
npm run start
```

**Reply by daniel-walz** [reply]
After som extra thoughts I have split the creation in two parts, so now I first build everything and then create a new container and copy the contents of the build folder to the new container and then run npm install there to get it more clean.

Dockerfile:
```
# base image for production
FROM mhart/alpine-node:14 as build

# SQLite3
RUN apk update
RUN apk add git > /dev/null

# set working directory
WORKDIR /app

# add environment variables
ENV PATH /app/node_modules/.bin:$PATH
ENV SKIP_PREFLIGHT_CHECK=true

# copy application files
COPY . .

# install and cache app development dependencies
RUN npm install --silent
RUN npm run build

# base image for production
FROM mhart/alpine-node:14

# SQLite3
RUN apk update
RUN apk add git sqlite python make g++ > /dev/null

# set working directory
WORKDIR /app

# copy application files
COPY --from=build /app/build /app
COPY --from=build /app/node_modules /app/node_modules
COPY ./scripts/start.sh /app

RUN mkdir tmp

# install and cache app production dependencies
RUN npm install sqlite3@5.0.0

# set startpoint for the application
RUN chmod +x start.sh
CMD ["/app/start.sh"]
```

start.sh:
```
#!/bin/sh

node ace migration:run --force
npm run start
```

**Reply by Nithur-M** [reply]
does this erase the db in every deployment? because you are creating the directory in every build right?

**Reply by rakis10** [reply]
thats what i looking for a long time, but yet to find solution

**Comment by deadcoder0904** [comment]
[I came to the same conclusion albeit a different stack.](https://stackoverflow.com/a/78034830/6141587)

**Reply by Nithur-M** [reply]
Hi, can i make this work without litestream? what should I look out for?
